File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* floyd(ListNode* head) {
        ListNode* slow = head;
        ListNode* fast = head;

        if(slow == nullptr || fast == nullptr) {
            return NULL;
        }
        while(fast != NULL && fast->next != NULL) {
            slow = slow->next;
            fast = fast->next->next;

            if(slow == fast) {
                return slow;
            }
        }
        return NULL;
    }
    ListNode *detectCycle(ListNode *head) {
        ListNode* intersection = floyd(head);

        if(intersection == NULL) {
            return NULL;
        }
        ListNode* slow = head;

        while(slow != intersection && intersection != NULL){
            slow = slow->next;
            intersection = intersection->next;
        }

        return slow;
    }
};

File: 0142-linked-list-cycle-ii/0142-linked-list-cycle-ii.java
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    static ListNode floyd(ListNode head) {
        ListNode slow = head;
        ListNode fast = head;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;

            if(slow == fast) {
                return slow;
            }
        }

        return null;
    }
    public ListNode detectCycle(ListNode head) {
        ListNode intersection = floyd(head);

        if(intersection == null) {
            return null;
        }
        ListNode start = head;
        while(start != intersection) {
            start = start.next;
            intersection = intersection.next;
        }
        return start;
    }
}